core/domain/state_domain.py

```diff
@@ -916,8 +938,946 @@ def get_all_outcomes(self) -> List[Outcome]:
+            )
+
+        try:
+            if len(self.answer_groups) > 0:
+                raise utils.ValidationError(
+                    'The Continue button does not require answer groups.'
+                )
+        except Exception:
+            pass
+
+    def _validate_end_exploration_input(self) -> None:
+        """Validates the End interaction.
+
+        Raises:
+            ValidationError. Raises a validation error in the following
+                condition:
+                - Recommended explorations are more than 3.
+        """
+        # Total recommended explorations should not be more than 3.
+        recc_exp_ids = (
+            self.customization_args['recommendedExplorationIds'].value
+        )
+        if len(recc_exp_ids) > 3:
+            raise utils.ValidationError(
+                'The End interaction should not have more than 3 '
+                'recommended explorations.'
+            )
+
+    def _set_lower_and_upper_bounds(
+        self,
+        range_var: RangeVariableDict,
+        lower_bound: float,
+        upper_bound: float,
+        lb_inclusive: bool,
+        ub_inclusive: bool
+    ) -> None:
+        """Helper function for NumericInput and FractionInput interaction,
+        Sets the lower and upper bounds for the range_var, mainly
+        we need to set the range so to keep track if any other rule's
+        range lies in between or not to prevent redundancy.
+
+        Args:
+            range_var: RangeVariableDict. To keep track of each rule's
+                ans group index, rule spec index, lower bound, upper bound,
+                lb inclusive, ub inclusive.
+            lower_bound: float. The lower bound.
+            upper_bound: float. The upper bound.
+            lb_inclusive: bool. If lower bound is inclusive.
+            ub_inclusive: bool. If upper bound is inclusive.
+        """
+        range_var['lower_bound'] = lower_bound
+        range_var['upper_bound'] = upper_bound
+        range_var['lb_inclusive'] = lb_inclusive
+        range_var['ub_inclusive'] = ub_inclusive
+
+    def _is_enclosed_by(
+        self,
+        test_range: RangeVariableDict,
+        base_range: RangeVariableDict
+    ) -> bool:
+        """Helper function for NumericInput and FractionInput interaction,
+        Returns `True` when `test_range` variable lies within
+        `base_range` variable.
+
+        Args:
+            test_range: RangeVariableDictDict. It represents the variable for
+                which we have to check the range. It keeps track of each rule's
+                ans group index, rule spec index, lower bound, upper bound,
+                lb inclusive, ub inclusive.
+            base_range: RangeVariableDictDict. It is the variable to which
+                the range is compared. It keeps track of other rule's
+                ans group index, rule spec index, lower bound, upper bound,
+                lb inclusive, ub inclusive.
```

Comment:
From these I'm not really sure which is whci. Also the "It keeps track of other rule's ans group index, rule spec index, lower bound, upper bound, lb inclusive, ub inclusive." is not needed.

Reply:
Done